network/network-address/src/lib.rs

```diff
@@ -263,12 +267,79 @@ impl NetworkAddress {
         Self(protocols)
     }
 
+    // TODO(philiphayes): could return NonZeroUsize?
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
     pub fn as_slice(&self) -> &[Protocol] {
         self.0.as_slice()
     }
 
-    pub fn push(&mut self, proto: Protocol) {
-        self.0.push(proto)
+    pub fn push(mut self, proto: Protocol) -> Self {
+        self.0.push(proto);
+        self
+    }
+
+    pub fn extend_from_slice(mut self, protos: &[Protocol]) -> Self {
+        self.0.extend_from_slice(protos);
+        self
+    }
+
+    /// Given a base `NetworkAddress`, append production protocols and
+    /// return the modified `NetworkAddress`.
+    ///
+    /// ### Example
+    ///
+    /// `/dns/example.com/tcp/6180` =>
+    /// `/dns/example.com/tcp/6180/ln-noise-ik/<network_pubkey>/ln-handshake/<handshake_version>`
+    // TODO(philiphayes): use handshake version enum
+    pub fn append_prod_protos(
+        self,
+        network_pubkey: x25519::PublicKey,
+        handshake_version: u8,
+    ) -> Self {
+        self.push(Protocol::NoiseIk(network_pubkey))
+            .push(Protocol::Handshake(handshake_version))
+    }
+
+    /// Given a base `NetworkAddress`, append production protocols and
+    /// return the modified `NetworkAddress`.
+    ///
+    /// ### Example
+    ///
+    /// `/ip4/127.0.0.1/tcp/6180` =>
+    /// `/ip4/127.0.0.1/tcp/6180/ln-peerid-ex/ln-handshake/<handshake_version>`
+    // TODO(philiphayes): gate with cfg(test)
+    // TODO(philiphayes): use handshake version enum
+    pub fn append_test_protos(self, handshake_version: u8) -> Self {
+        self.push(Protocol::PeerIdExchange)
+            .push(Protocol::Handshake(handshake_version))
+    }
+
+    /// A temporary, hacky function to parse out the first `/ln-noise-ik/<pubkey>` from
+    /// a `NetworkAddress`. We can remove this soon, when we move to the interim
+    /// "monolithic" transport model.
+    pub fn find_noise_proto(&self) -> Option<&x25519::PublicKey> {
+        self.0.iter().find_map(|proto| match proto {
+            Protocol::NoiseIk(pubkey) => Some(pubkey),
```

Comment:
`x25519::PublicKey` are `Copy` so you should be able to return a value not a reference :o

Reply:
what's the "monolithic" transport model?

Reply:
been sketching this out... probably will look something like this 

```rust
impl MemoryNoiseMutualAuthTransport {
    pub fn new(
        identity_key: x25519::PrivateKey,
        trusted_peers: TrustedPeersMap,
        application_protocols: SupportedProtocols,
    ) -> Self {
        let mut own_handshake = HandshakeMsg::new();
        own_handshake.add(SUPPORTED_MESSAGING_PROTOCOL, application_protocols);

        Self {
            noise_config: Arc::new(NoiseConfig::new(identity_key)),
            own_handshake,
            trusted_peers,
        }
    }

    pub async fn dial(
        &self,
        addr: NetworkAddress,
    ) -> io::Result<Connection<NoiseSocket<MemorySocket>>> {
        use libra_network_address::Protocol::*;
        let origin = ConnectionOrigin::Outbound;

        let (port, pubkey, version) = match addr.as_slice() {
            [Memory(port), NoiseIk(pubkey), Handshake(version)] => (*port, *pubkey, *version),
            _ => {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!(
                        "Unexpected NetworkAddress: '{}', expected: \
                         '/memory/<port>/ln-noise-ik/<pubkey>/ln-handshake/<version>'",
                        addr
                    ),
                ))
            }
        };
        verify_handshake_version_supported(version)?;

        let socket = MemorySocket::connect(port)?;

        let (remote_static_key, socket) =
            self.noise_config.upgrade_connection(socket, origin).await?;
        verify_pubkeys_match(&remote_static_key, &pubkey)?;

        let peer_id = identity_key_to_peer_id(&self.trusted_peers, &remote_static_key)
            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "Not a trusted peer"))?;

        perform_handshake(peer_id, socket, addr, origin, &self.own_handshake).await
    }

    pub fn listen_on(&self, addr: NetworkAddress) -> io::Result<()> {
        use libra_network_address::Protocol::*;

        let origin = ConnectionOrigin::Inbound;

        let port = match addr.as_slice() {
            [Memory(port)] => *port,
            _ => {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!(
                        "Unexpected NetworkAddress: '{}', expected: '/memory/<port>'",
                        addr
                    ),
                ))
            }
        };

        let mut listener = MemoryListener::bind(port)?;
        let port = listener.local_addr();

        Ok(listener.incoming().and_then(|socket| async move {
            let (remote_static_key, socket) =
                self.noise_config.upgrade_connection(socket, origin).await?;
            verify_pubkeys_match(&remote_static_key, &pubkey)?;

            let peer_id = identity_key_to_peer_id(&self.trusted_peers, &remote_static_key)
                .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "Not a trusted peer"))?;

            perform_handshake(peer_id, socket, addr, origin, &self.own_handshake).await
        }))

        todo!()
    }
}
```

Reply:
ah yes! You already explained that to me :) I'm all for it!

Reply:
let me get my noise PR in first though :P (I already verify public keys in the noise upgrade_connection itself)